src/http2.rs

```diff
@@ -0,0 +1,563 @@
+    }
+}
+
+/// The `Protocol` implementation that provides HTTP/2 messages (i.e. `Http2Message`).
+pub struct Http2Protocol<C, S> where C: NetworkConnector<Stream=S> + Send + 'static,
+                                 S: NetworkStream + Send + Clone {
+    connector: C,
+}
+
+impl<C, S> Http2Protocol<C, S> where C: NetworkConnector<Stream=S> + Send + 'static,
+                                     S: NetworkStream + Send + Clone {
+    /// Create a new `Http2Protocol` that will use the given `NetworkConnector` to establish TCP
+    /// connections to the server.
+    pub fn with_connector(connector: C) -> Http2Protocol<C, S> {
+        Http2Protocol {
+            connector: connector,
+        }
+    }
+
+    /// A private helper method that creates a new `SimpleClient` that will use the given
+    /// `NetworkStream` to communicate to the remote host.
+    fn new_client(&self, stream: S, host: String, scheme: HttpScheme)
+            -> ::Result<SimpleClient<Http2Stream<S>>> {
+        Ok(try!(SimpleClient::with_connector(Http2Connector {
+            stream: stream,
+            scheme: scheme,
+            host: host,
+        })))
+    }
+}
+
+impl<C, S> Protocol for Http2Protocol<C, S> where C: NetworkConnector<Stream=S> + Send + 'static,
+                                                  S: NetworkStream + Send + Clone {
+    fn new_message(&self, host: &str, port: u16, scheme: &str) -> ::Result<Box<HttpMessage>> {
+        let stream = try!(self.connector.connect(host, port, scheme)).into();
+
+        let scheme = match scheme {
+            "http" => HttpScheme::Http,
+            "https" => HttpScheme::Https,
+            _ => return Err(From::from(Http2Error::from(
+                        io::Error::new(io::ErrorKind::Other, "Invalid scheme")))),
+        };
+        let client = try!(self.new_client(stream, host.into(), scheme));
+
+        Ok(Box::new(Http2Message::with_client(client)))
+    }
+
+    #[inline]
+    fn set_ssl_verifier(&mut self, verifier: ContextVerifier) {
+        self.connector.set_ssl_verifier(verifier)
+    }
+}
+
+/// Represents an HTTP/2 request, described by a `RequestHead` and the body of the request.
+/// A convenience struct only in use by the `Http2Message`.
+#[derive(Clone, Debug)]
+struct Http2Request {
+    head: RequestHead,
+    body: Vec<u8>,
+}
+
+/// Represents an HTTP/2 response.
+/// A convenience struct only in use by the `Http2Message`.
+#[derive(Clone, Debug)]
+struct Http2Response {
+    body: Cursor<Vec<u8>>,
+}
+
+/// An implementation of the `HttpMessage` trait for HTTP/2.
+///
+/// Relies on the `solicit::http::SimpleClient` for HTTP/2 communication. Adapts both outgoing and
+/// incoming messages to the API that `hyper` expects in order to be able to use the message in
+/// the `hyper::client` module.
+pub struct Http2Message<S> where S: CloneableStream {
+    client: SimpleClient<Http2Stream<S>>,
+    request: Option<Http2Request>,
```

Comment:
Here, on the other hand, the state enum works out just fine. I suppose because there's no state that needs to be transferred between the variants when a transition happens...